### PR TITLE
*: implement a retry logic for auth old revision in the client

### PR DESCRIPTION
**Pull request by mitake**
fix https://github.com/etcd-io/etcd/issues/13300

@davissp14 could you try this one? I tried the change with https://github.com/mitake/etcd-things/tree/master/auth-failover and it worked

- [x] add test, ideally for retry_intercepter.go

### CODE REVIEW

File: client/v3/retry_interceptor.go [hunk 1]
```diff
@@ -157,7 +157,8 @@ func (c *Client) shouldRefreshToken(err error, callOpts *options) bool {
 		// which is possible when the client token is cleared somehow
 		return c.authTokenBundle != nil // equal to c.Username != "" && c.Password != ""
 	}
-	return callOpts.retryAuth && rpctypes.Error(err) == rpctypes.ErrInvalidAuthToken
+
+	return callOpts.retryAuth && (errors.Is(err, rpctypes.ErrInvalidAuthToken) || errors.Is(err, rpctypes.ErrGRPCAuthOldRevision))
```
Comment (mitake) [inline client/v3/retry_interceptor.go:161]: @r-ashish could you check this path? I think we should use `errors.Is()` instead of `==` here.

Comment (mitake) [reply client/v3/retry_interceptor.go:161]: also line 155 would need change too. I'll add a unit test to the file for ensuring that the error handling logic is correct

Comment (r-ashish) [reply client/v3/retry_interceptor.go:161]: yeah agree, looks okay.

**Comment by davissp14** [comment]
@mitake Seems to recover nicely!

**Comment by davissp14** [comment]
Actually, one thing I did find is that tokens are not renewed at expire time:

```
{"level":"warn","ts":"2021-08-30T08:45:13.387-0500","logger":"etcd-client","caller":"v3/retry_interceptor.go:63","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00000c780/#initially=[http://127.0.0.1:2379]","attempt":0,"error":"rpc error: code = Unauthenticated desc = etcdserver: invalid auth token"}
Get failed: etcdserver: invalid auth token
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x160a32a]

goroutine 1 [running]:
main.main()
	/Users/davissp14/go/src/github.com/fly-examples/etcd/test.go:37 +0x48a
exit status 2
```

**Comment by mitake** [comment]
@davissp14 thanks for checking! I’ll fix the renew issue. Probably I can finalize this PR this weekend (or sometime next week), is this ok for you?

**Comment by mitake** [comment]
@davissp14 the issue should be fixed now, could you try again?

**Comment by mitake** [comment]
I think the change is ready to be reviewed, could you take a look @spzala @serathius ?

**Review by spzala — APPROVED** [review]
@mitake lgtm overall. Great job adding test :) Thanks!

**Comment by mitake** [comment]
@spzala thanks! I'm merging this PR then.

**Comment by mitake** [comment]
@davissp14 did this MR solve the issue on your env? I'll also make PRs for backporting this PR (and related ones) to stable releases.

**Pull request merged**